### PR TITLE
v2.0.2 plan

### PR DESCRIPTION
**Pull request by ragaeeb**
# v2.0.2 Plan: Aggressive P1+P2 Stability, Refactor, and Debt Removal

## Summary
- Scope is `P1 + P2` in one release, with strict TDD slices and strict gates.
- Baseline is currently green (`bun test`, `bun run tsc --noEmit`, `bun run build` all pass on current `HEAD`).
- We will implement correctness hardening first, then deep decomposition (runner first), then hygiene passes.
- We will drop legacy protocol compatibility now (breaking internal compatibility is allowed).

## Locked Decisions Applied
1. Aggressive scope: include both `P1` and `P2` items from `docs/reviews/synthesis.md`.
2. TDD slices only: test-first for each slice, then minimal implementation/refactor.
3. No feature flags.
4. Deep decomposition starts with `utils/platform-runner.ts`.
5. Lease strategy for this release: harden current lease + tests now; stage storage/arbitration migration for next release.
6. Legacy compatibility debt: remove now (no backward-compat shim for legacy attempt-less lifecycle messages).

## Reviewer Source Protocol (mandatory per slice)
Use these before coding each slice:
1. `docs/reviews/synthesis.md` (fix IDs, adjudicated direction).
2. `docs/reviews/opus-4.6/claude-opus-4.6-implementation-guide.md` (decomposition sequence and extraction boundaries).
3. `docs/reviews/qwen-3.5-plus.md` and `docs/reviews/glm-5.md` (retry/lease/lifecycle race patterns).
4. `docs/reviews/gpt-5.2-thinking.md` (bounded caches, title centralization, architecture pressure).
5. `docs/reviews/coderabbit.md` (targeted bug-level assertions in touched files).

## Public APIs / Interfaces / Types Changes
1. `utils/protocol/messages.ts`:
- Remove `LegacyLifecycleMessage`, `LegacyFinishedMessage`, `LegacyStreamDeltaMessage`.
- Remove `isLegacyLifecycleMessage`, `isLegacyFinishedMessage`, `isLegacyStreamDeltaMessage`, `buildLegacyAttemptId`.
- Keep `attemptId` mandatory for lifecycle/finished/delta wire contracts.

2. `platforms/types.ts`:
- Add optional `isPlatformGenerating?(): boolean` to `LLMPlatform`.
- Runner will call adapter-provided generator detection instead of per-platform inline DOM heuristics.

3. New shared title policy surface (internal but exported):
- Add `utils/title-resolver.ts` with deterministic precedence and generic-title classification utilities used by runner, interception manager, and adapters.

4. Decomposition interfaces (internal):
- Add `utils/runner/*` modules with stable `runPlatform()` re-export preserved at `utils/platform-runner.ts`.
- Add `entrypoints/interceptor/*` modules while keeping `entrypoints/interceptor.content.ts` as thin entrypoint.

## TDD Implementation Slices

1. Slice 0: Baseline + traceability setup
- Tests first: none (baseline capture only).
- Commands: `bun test`, `bun run tsc --noEmit`, `bun run build`.
- Record in working notes per slice: fix ID(s), reviewer source file, tests added first.

2. Slice 1: Protocol debt removal (legacy compat deletion)
- Files: `utils/protocol/messages.ts`, `utils/protocol/messages.test.ts`, `integration/legacy-message-compat.test.ts`.
- Tests first:
  - Add strict-message tests asserting attempt-less lifecycle/finished/delta are invalid.
  - Replace/remove legacy integration test with strict attempt-aware protocol integration assertions.
- Implementation:
  - Delete legacy types/helpers/guards.
  - Tighten tests and imports accordingly.
- Acceptance:
  - No legacy helpers remain in tree.
  - All protocol and integration suites pass under strict attempt-aware contract.

3. Slice 2: A-07 stabilization retry monotonic concurrency hardening
- Files: `utils/platform-runner.ts`, `utils/platform-runner.test.ts`, `integration/navigation-dispose-sequence.test.ts`, `integration/dispose-retry-race.test.ts`.
- Tests first:
  - No concurrent stabilization tick for same attempt.
  - Retry callbacks no-op after dispose/navigation.
  - All retry state maps clear together on cancellation.
- Implementation:
  - Add per-attempt in-progress guard for retry tick.
  - Re-check `isAttemptDisposedOrSuperseded` after await boundaries in retry flow.
  - Enforce atomic cleanup across retry timer/count/start/timeout-warning maps.
- Acceptance:
  - No stale retry work runs after disposal.
  - Retry scheduling remains deterministic under multi-tab race tests.

4. Slice 3: A-08 title resolver centralization
- Files: `utils/title-resolver.ts` (new), `utils/title-resolver.test.ts` (new), `utils/platform-runner.ts`, `utils/managers/interception-manager.ts`, `platforms/gemini.ts`, `utils/gemini-stream-parser.ts`.
- Tests first:
  - Cross-platform generic/specific title matrix in `title-resolver.test.ts`.
  - Integration precedence test: stream > specific-cache > DOM > first-user fallback.
  - Late generic title must not clobber earlier specific title.
- Implementation:
  - Replace duplicated generic-title checks with shared policy.
  - Ensure interception cache stickiness remains invariant.
  - Ensure export-time resolution consumes shared resolver.
- Acceptance:
  - All title decisions flow through shared module.
  - Gemini/Grok/runner/interception behavior matches precedence invariants.

5. Slice 4: A-09 interceptor bounded cache unification + cleanup discipline
- Files: `entrypoints/interceptor.content.ts`, `utils/lru-cache.ts` (if needed), `tests/interceptor.content.test.ts` (expanded or split).
- Tests first:
  - Bound/evict behavior for dedupe caches/maps/sets.
  - Attempt disposal clears attempt-scoped cache entries.
  - Long-session churn does not grow unbounded.
- Implementation:
  - Replace raw unbounded maps/sets with bounded structures and periodic prune where TTL-based.
  - Add deterministic cleanup hooks on attempt disposal/teardown.
- Acceptance:
  - No unbounded interceptor cache paths remain for runtime-critical maps/sets.
  - Existing behavior remains unchanged except bounded memory.

6. Slice 5: A-13 parser edge-case hardening (tests-first, minimal logic change)
- Files: `platforms/gemini.ts`, `platforms/gemini.test.ts`, `utils/gemini-stream-parser.ts`, `utils/gemini-stream-parser.test.ts`, `utils/grok-stream-parser.ts`, `utils/grok-stream-parser.test.ts`, `platforms/grok.test.ts`.
- Tests first:
  - Gemini envelope/index drift fixtures.
  - Grok NDJSON split-line/chunk-boundary fixtures.
  - Malformed chunk continuation and non-throw assertions.
- Implementation:
  - Apply minimal parse heuristics to satisfy failing edge tests only.
- Acceptance:
  - No uncaught parser crashes in new fixtures.
  - Existing parser behavior stays stable outside covered gaps.

7. Slice 6: A-11 current lease hardening (no storage-channel migration yet)
- Files: `utils/sfe/cross-tab-probe-lease.ts`, `utils/sfe/cross-tab-probe-lease.test.ts`, `integration/probe-lease-collision-expiry.test.ts`, `utils/platform-runner.ts`.
- Tests first:
  - Collision/write-verify race scenario coverage.
  - Expiry handoff with active retries remains single-owner.
  - Release semantics on navigation/dispose paths are complete.
- Implementation:
  - Tighten claim/verify/retry semantics and release-path coverage.
  - Keep existing localStorage lease architecture this release.
- Acceptance:
  - Deterministic single-owner probing in integration tests.
  - No orphan retry probes after owner disposal.

8. Slice 7: A-12 utility dedup consolidation
- Files: `utils/text-utils.ts`, `utils/text-utils.test.ts`, plus touched adapter/runner/interceptor files.
- Tests first:
  - Characterization tests for duplicated helper behaviors before extraction.
- Implementation:
  - Consolidate duplicated text dedupe/truncation/history helpers.
  - Consolidate filename/date fallback helpers where duplicated.
- Acceptance:
  - Shared utilities replace duplicates without behavior drift.

9. Slice 8: A-15 runner decomposition phase 1 (runner-first)
- Files new: `utils/runner/state.ts`, `utils/runner/dom-snapshot.ts`, `utils/runner/export-pipeline.ts`.
- Tests first:
  - Add characterization tests around snapshot build and export/title behavior before move.
- Implementation:
  - Introduce `RunnerState`.
  - Move DOM snapshot + export pipeline logic with no behavior changes.
  - Keep `utils/platform-runner.ts` exporting `runPlatform` to preserve call sites.
- Acceptance:
  - Moved code paths are behavior-identical (characterization tests unchanged).

10. Slice 9: A-15 runner decomposition phase 2 (message bridge, lifecycle, stream probe, calibration)
- Files new: `utils/runner/message-bridge.ts`, `utils/runner/lifecycle-manager.ts`, `utils/runner/stream-probe.ts`, `utils/runner/calibration-runner.ts`, `utils/runner/index.ts`.
- Tests first:
  - Harness tests around event ordering, lifecycle transitions, and bridge request/response.
- Implementation:
  - Extract subsystem-by-subsystem, one commit per subsystem.
  - Keep logic unchanged during moves; behavioral changes only in dedicated follow-up commits.
- Acceptance:
  - `utils/platform-runner.ts` becomes thin wrapper.
  - All existing runner/integration tests stay green through each extraction step.

11. Slice 10: A-15 interceptor decomposition (after runner split is stable)
- Files new: `entrypoints/interceptor/stream-monitors/chatgpt-sse-monitor.ts`, `entrypoints/interceptor/stream-monitors/gemini-stream-monitor.ts`, `entrypoints/interceptor/stream-monitors/grok-stream-monitor.ts`, `entrypoints/interceptor/fetch-wrapper.ts`, `entrypoints/interceptor/xhr-wrapper.ts`, `entrypoints/interceptor/proactive-fetcher.ts`.
- Tests first:
  - Unit tests for extracted monitors and wrappers.
- Implementation:
  - Move logic from monolith into modules with unchanged behavior.
  - Keep `entrypoints/interceptor.content.ts` as thin wiring entrypoint.
- Acceptance:
  - Entry point is thin and deterministic.
  - Extracted units have direct tests (closing current interceptor unit-test gap).

12. Slice 11: A-14 logging hygiene + A-16 type-safety consistency
- Files: `utils/minimal-logs.ts`, `utils/minimal-logs.test.ts`, `utils/platform-runner.ts` and extracted runner modules, `entrypoints/interceptor.content.ts` and extracted interceptor modules.
- Tests first:
  - Log dedupe/noise-level assertions.
  - Type guard tests where `unknown` narrows replace `any`.
- Implementation:
  - Normalize parse/log levels for noisy paths.
  - Add optional log transport batching only if measurable channel pressure appears in profiling.
  - Incrementally replace hot-path `any` with `unknown`+guards.
- Acceptance:
  - Reduced log noise without losing diagnostics.
  - `tsc --noEmit` remains strict green.

13. Slice 12: Docs + backlog status closure for v2.0.2
- Files: `docs/reviews/synthesis.md`, `docs/architecture.md`, `AGENTS.md`, `README.md` (only if stale).
- Required updates:
  - Mark completed P1/P2 items in synthesis with date and evidence.
  - Add explicit remaining backlog section (post-v2.0.2) including lease storage/arbitration migration and deferred security hardening.
  - Update architecture for new module boundaries and title/lease/protocol invariants.
- Acceptance:
  - Documentation matches implemented architecture and remaining roadmap.

## Validation Gates and Smoke Scenarios
1. Per-slice TDD gate:
- New tests fail before fix/refactor.
- Targeted suite passes after change.
- `bun run tsc --noEmit` passes.

2. Decomposition slices gate:
- `bun run build` passes after each major extraction slice.

3. End-of-wave mandatory gate:
- `bun test`
- `bun run tsc --noEmit`
- `bun run build`

4. High-risk suites rerun at end:
- `utils/platform-runner.test.ts`
- `tests/interceptor.content.test.ts`
- `platforms/gemini.test.ts`
- `platforms/grok.test.ts`
- `utils/sfe/cross-tab-probe-lease.test.ts`
- `integration/navigation-dispose-sequence.test.ts`
- `integration/probe-lease-collision-expiry.test.ts`

5. Manual smoke checkpoints:
- After Slice 6 (all correctness hardening): ChatGPT/Gemini/Grok multi-tab streaming + Save gating + title fidelity.
- After Slice 12 (release candidate): repeat full 3-platform smoke with 4+ tabs each and long reasoning sessions.

## Explicit Remaining Backlog After v2.0.2
1. Lease architecture migration to extension-scoped arbitration (`chrome.storage.session`/background coordinator) is planned next release.
2. Deferred security hardening items from synthesis remain out of scope for v2.0.2.
3. Any optional event-bus/state-container rewrite remains deferred until after decomposition stabilizes in production.

## Assumptions and Defaults
1. Breaking internal protocol compatibility is acceptable in v2.0.2.
2. No external/public consumers depend on legacy attempt-less wire payloads.
3. Runtime behavior changes are limited to adjudicated P1/P2 goals; no new user-facing features.
4. Refactor slices prioritize behavior-preserving extraction first; logic changes happen only in dedicated, test-backed slices.

---

**Continue Tasks:** ✅ 3 no changes — [View all](https://hub.continue.dev/inbox/pr/ragaeeb/blackiya/22?utm_source=github_pr&utm_medium=pr_body&utm_campaign=continue_tasks)

## Summary by CodeRabbit

* **New Features**
  * Modular runner/interceptor architecture, new fetch/XHR interception helpers, stream monitors, proactive fetch gating, title-resolution/export utilities, runner state & snapshot builders, and platform generation-state checks.

* **Bug Fixes**
  * Tightened message protocol (attemptId required), removed legacy compatibility helpers, improved cross-tab lease robustness, safer logging/context handling, and improved stream/NDJSON parsing.

* **Documentation**
  * README and architecture docs updated to reflect new structure.

* **Tests**
  * Expanded unit and integration coverage.

* **Version**
  * Bumped to v2.0.2

### CODE REVIEW

File: entrypoints/interceptor/proactive-fetcher.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+export class ProactiveFetcher {
+    private readonly inFlight = new Set<string>();
+
+    public markInFlight(key: string): boolean {
+        if (this.inFlight.has(key)) {
+            return false;
+        }
+        this.inFlight.add(key);
+        return true;
+    }
```
[the remark this answers is not in the document]
Comment (ragaeeb) [reply entrypoints/interceptor/proactive-fetcher.ts]: https://github.com/ragaeeb/blackiya/pull/22/commits/316ca3b6881acfaa0c277b72518260b179f3e907
